server/src/test/java/org/elasticsearch/action/admin/indices/refresh/RefreshResponseTests.java

```diff
@@ -0,0 +1,113 @@
+
+public class RefreshResponseTests extends ESTestCase {
+
+    private static List<DefaultShardOperationFailedException> failures = new ArrayList<>();
+
+    private static class FakeElasticsearchException extends ElasticsearchException {
+
+        private Index index;
+        private ShardId shardId;
+        private RestStatus status;
+
+        public FakeElasticsearchException(String index, int shardId, RestStatus status, String msg) {
+            super(msg);
+            this.index = new Index(index, "_na_");
+            this.shardId = new ShardId(this.index, shardId);
+            this.status = status;
+        }
+
+        @Override
+        public Index getIndex() {
+            return this.index;
+        }
+
+        @Override
+        public ShardId getShardId() {
+            return this.shardId;
+        }
+
+        @Override
+        public RestStatus status() {
+            return this.status;
+        }
+    }
+
+    @BeforeClass
+    public static void prepareException() {
+        failures = new ArrayList<>();
+        failures.add(new DefaultShardOperationFailedException(
+            new FakeElasticsearchException("index1", 1, RestStatus.INTERNAL_SERVER_ERROR, "fake exception 1")));
+        failures.add(new DefaultShardOperationFailedException(
+            new FakeElasticsearchException("index2", 2, RestStatus.GATEWAY_TIMEOUT, "fake exception 2")));
+    }
+
+    public void testToXContent() {
+        RefreshResponse response = new RefreshResponse(10, 10, 0, null);
+        String output = Strings.toString(response);
+        assertEquals("{\"_shards\":{\"total\":10,\"successful\":10,\"failed\":0}}", output);
+    }
+
+    public void testToAndFromXContent() throws IOException {
+        doFromXContentTestWithRandomFields(false);
+    }
+
+    private void doFromXContentTestWithRandomFields(boolean addRandomFields) throws IOException {
+        RefreshResponse response = new RefreshResponse(10, 10, 0, failures);
+        boolean humanReadable = randomBoolean();
+        XContentType xContentType = randomFrom(XContentType.values());
+        BytesReference bytesReference = toShuffledXContent(response, xContentType, ToXContent.EMPTY_PARAMS, humanReadable);
+        if (addRandomFields) {
+            bytesReference = insertRandomFields(xContentType, bytesReference, null, random());
+        }
+        RefreshResponse parsedResponse;
+        try(XContentParser parser = createParser(xContentType.xContent(), bytesReference)) {
+            parsedResponse = RefreshResponse.fromXContent(parser);
+            assertNull(parser.nextToken());
+        }
+
+        assertThat(response.getTotalShards(), equalTo(parsedResponse.getTotalShards()));
+        assertThat(response.getSuccessfulShards(), equalTo(parsedResponse.getSuccessfulShards()));
+        assertThat(response.getFailedShards(), equalTo(parsedResponse.getFailedShards()));
+        assertThat(response.getShardFailures(), equalTo(parsedResponse.getShardFailures()));
```

Comment:
I am afraid that this assertion may not work, given that what we parse back is always an ElasticsearchException rather than the original type. That is expected and makes testing exceptions parsing trickier. See other tests around this.